File: occasum-core/src/main/java/pw/stamina/occasum/properties/enums/EnumParsingService.java
package pw.stamina.occasum.properties.enums;

import pw.stamina.occasum.properties.PropertyParseException;

public interface EnumParsingService<T extends Enum<T>> {

    T parse(String input) throws PropertyParseException;

    String toString(T value);

    T[] getEnumConstants();

    static <T extends Enum<T>> EnumParsingService<T> standard(Class<T> enumClass) {
        return new StandardEnumParsingService<>(enumClass);
    }

    static <T extends Enum<T>> EnumParsingService<T> caseInsensitive(Class<T> enumClass) {
        return new CaseInsensitiveEnumParsingService<>(enumClass);
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/properties/enums/EnumParsingService.kt
package pw.stamina.occasum.properties.enums

import pw.stamina.occasum.properties.PropertyParseException

interface EnumParsingService<T : Enum<T>> {

    val enumConstants: Array<out T>

    @Throws(PropertyParseException::class)
    fun parse(input: String): T

    fun toString(value: T): String

    companion object {

        fun <T : Enum<T>> standard(enumClass: Class<T>): EnumParsingService<T> {
            return StandardEnumParsingService(enumClass)
        }

        fun <T : Enum<T>> caseInsensitive(enumClass: Class<T>): EnumParsingService<T> {
            return CaseInsensitiveEnumParsingService(enumClass)
        }
    }
}
